src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -6257,186 +6257,195 @@ void StorageReplicatedMergeTree::replacePartitionFrom(
+                entry_replace.part_names_checksums.emplace_back(checksum);
+            entry_replace.columns_version = -1;
+        }
 
-    ReplicatedMergeTreeLogEntryData entry;
-    {
-        auto src_table_id = src_data.getStorageID();
-        entry.type = ReplicatedMergeTreeLogEntryData::REPLACE_RANGE;
-        entry.source_replica = replica_name;
-        entry.create_time = time(nullptr);
-        entry.replace_range_entry = std::make_shared<ReplicatedMergeTreeLogEntryData::ReplaceRangeEntry>();
-
-        auto & entry_replace = *entry.replace_range_entry;
-        entry_replace.drop_range_part_name = drop_range_fake_part_name;
-        entry_replace.from_database = src_table_id.database_name;
-        entry_replace.from_table = src_table_id.table_name;
-        for (const auto & part : src_parts)
-            entry_replace.src_part_names.emplace_back(part->name);
-        for (const auto & part : dst_parts)
-            entry_replace.new_part_names.emplace_back(part->name);
-        for (const String & checksum : part_checksums)
-            entry_replace.part_names_checksums.emplace_back(checksum);
-        entry_replace.columns_version = -1;
-    }
-
-    /// Remove deduplication block_ids of replacing parts
-    if (replace)
-        clearBlocksInPartition(*zookeeper, drop_range.partition_id, drop_range.max_block, drop_range.max_block);
+        /// Remove deduplication block_ids of replacing parts
+        if (replace)
+            clearBlocksInPartition(*zookeeper, drop_range.partition_id, drop_range.max_block, drop_range.max_block);
 
-    DataPartsVector parts_to_remove;
-    Coordination::Responses op_results;
+        DataPartsVector parts_to_remove;
+        Coordination::Responses op_results;
 
-    try
-    {
-        Coordination::Requests ops;
-        for (size_t i = 0; i < dst_parts.size(); ++i)
+        try
         {
-            getCommitPartOps(ops, dst_parts[i], block_id_paths[i]);
-            ephemeral_locks[i].getUnlockOps(ops);
-
-            if (ops.size() > zkutil::MULTI_BATCH_SIZE)
+            Coordination::Requests ops;
+            for (size_t i = 0; i < dst_parts.size(); ++i)
             {
-                /// It is unnecessary to add parts to working set until we commit log entry
-                zookeeper->multi(ops);
-                ops.clear();
+                getCommitPartOps(ops, dst_parts[i], block_id_paths[i]);
+                ephemeral_locks[i].getUnlockOps(ops);
+
+                if (ops.size() > zkutil::MULTI_BATCH_SIZE)
+                {
+                    /// It is unnecessary to add parts to working set until we commit log entry
+                    zookeeper->multi(ops);
+                    ops.clear();
+                }
             }
-        }
 
-        if (auto txn = query_context->getZooKeeperMetadataTransaction())
-            txn->moveOpsTo(ops);
+            if (auto txn = query_context->getZooKeeperMetadataTransaction())
+                txn->moveOpsTo(ops);
```

Comment:
It will not work correctly with retries

Reply:
I'm not sure how this `ZooKeeperMetadataTransaction` works in general. But it should provide a way to allow retries.

Reply:
I think we can just disable retries for `DatabaseReplicated` (like [here](https://github.com/ClickHouse/ClickHouse/blob/92eec74ad7edc1be3cbee52bbb1e4946dd7812b3/src/Storages/StorageReplicatedMergeTree.cpp#L4800-L4804)), because it has own retry mechanism

Reply:
Done.